components/amara/sources/common.mjs

```diff
@@ -0,0 +1,102 @@
+import amara from "../amara.app.mjs";
+import constants from "../constants.mjs";
+
+export default {
+  props: {
+    amara,
+    db: "$.service.db",
+    timer: {
+      type: "$.interface.timer",
+      label: "Polling schedule",
+      description: "How often to poll the Amara API",
+      default: {
+        intervalSeconds: 60 * 15,
+      },
+    },
+    team: {
+      description: "The team slug of the team to watch",
+      optional: true,
+      propDefinition: [
+        amara,
+        "team",
+      ],
+    },
+  },
+  methods: {
+    setLastUrl(lastUrl) {
+      this.db.set(constants.LAST_URL, lastUrl);
+    },
+    getLastUrl() {
+      return this.db.get(constants.LAST_URL);
+    },
+    getResourceFn() {
+      throw new Error("getResourceFn is not implemented");
+    },
+    getResourceFnArgs() {
+      throw new Error("getResourceFnArgs is not implemented");
+    },
+    getAllowedEvents() {
+      throw new Error("getAllowedEvents is not implemented");
+    },
+    /**
+     * For Team notifications returns resource.data.event and
+     * resource.data.number as the identifier.
+     *  - For response data look at the docs [here](https://apidocs.amara.org/#list-team-notifications)
+     * For Team activity returns resource.type and resource.video as the identifier.
+     *  - For response data look at the docs [here](https://apidocs.amara.org/#team-activity)
+     * @param {object} resource
+     * @returns {Array}
+     */
+    getEventTypeAndId(resource) {
+      if (resource.data) {
+        return [
+          resource.data.event,
+          resource.data.number,
+        ];
+      }
+      return [
+        resource.type,
+        resource.title || resource.video,
+      ];
+    },
+    generateMeta(resource) {
+      const [
+        eventType,
+        eventId,
+      ] = this.getEventTypeAndId(resource);
+      return {
+        id: JSON.stringify(resource),
+        summary: `${eventId} ${eventType}`,
+        ts: Date.parse(resource.data ?? resource.timestamp ?? Date.now()),
+      };
+    },
+    isRelevant(resource) {
+      const [
+        eventType,
+      ] = this.getEventTypeAndId(resource);
+      const allowedEvents = this.getAllowedEvents();
+      return allowedEvents.includes(eventType);
+    },
+    processEvent(resource) {
+      if (this.isRelevant(resource)) {
+        const meta = this.generateMeta(resource);
+        this.$emit(resource, meta);
+      }
+    },
+  },
+  async run({ $ }) {
+    const url = this.getLastUrl();
+
+    const { lastUrl } = await this.amara.paginateResources({
+      resourceFn: this.getResourceFn(),
+      resourceFnArgs: this.getResourceFnArgs({
+        $,
+        url,
+      }),
+      limit: 40,
+      callback: this.processEvent,
+    });
+
+    this.setLastUrl(lastUrl);
```

Comment:
If more than the [default max](https://github.com/PipedreamHQ/pipedream/blob/7463b1b9957cd0edf5d75e9951b79a300a5b05d1/components/amara/constants.mjs#L7) of 60 event resources have been created in Amara, `lastUrl` will have an offset of 60 at `this.setLastUrl(lastUrl)`. If I'm not mistaken, on the next `run` (and following runs), the source may fetch and `$emit` duplicate events that have not been emitted in the last [100 cached events](https://pipedream.com/docs/components/api/#dedupe-strategies).

I think we could fix this issue by removing `url` from the `resourceFnArgs`, but events would be missed if there are more than `DEFAULT_MAX_ITEMS` new events between source runs.

Preferably, we could use an approach similar to in [Textlocal sources](https://github.com/PipedreamHQ/pipedream/blob/3f6545d06638d674c6b15b06a036abc48de67a8d/components/textlocal/sources/new-sent-api-message/new-sent-api-message.js#L40-L41) where we paginate items [until the last](https://github.com/PipedreamHQ/pipedream/blob/3f6545d06638d674c6b15b06a036abc48de67a8d/components/textlocal/textlocal.app.js#L93) emitted item (ID or timestamp) in the previous run.
What do you think?

Reply:
That's right but I think the problem is because I'm setting `null` into `lastUrl` at the end of the `run` so I can fix this by just doing
```javascript
        nextResources = nextResponse.objects;
        resources = resources.concat(nextResources);

        if (nextResponse.meta.next) {
          lastUrl = nextResponse.meta.next;
        }

        if (callback) {
          nextResources.forEach(callback);
        }

      } while (nextResponse?.meta.next && resources.length < max);
```

Reply:
It looks like [Notifications](https://apidocs.amara.org/#list-team-notifications) and [Activity](https://apidocs.amara.org/#team-activity) are in reverse creation order (descending `date`/`timestamp`), so I think we would be fetching older resources on each request if we remove setting `lastUrl` to `null`.

<img width="300" alt="timestamp" src="https://user-images.githubusercontent.com/19861096/144676666-792d749a-1472-4ef2-a16b-78c2963fc632.png">


The [Team activity](https://apidocs.amara.org/#team-activity) endpoint has an `after` query parameter, which we could use to include only activity after the most recent activity event emitted. But [Notifications](https://apidocs.amara.org/#list-team-notifications) doesn't seem to have an `after` or `sort` query parameter.

Reply:
Hi @js07  It seems like is not even sorting in descending order but ascending order contrary to what the docs say. So I did something similar to what you pointed out in the previous comment but without taking into account the `id` but the whole object as the identifier since there is no unique property I can use here. I'm going to wait for an actual account that has an Enterprise plan so I can test with real data the rest of notifications events. @dylburger is going to be helping me out on that.

Reply:
I have an Enterprise account so if that's useful for helping move anything along quicker, I'm happy to assist where I'm able.  There's probably no substitute for seeing the real data in your own environment, but if there's utility in me helping by using our account, let me know what you'd like me to do.